lib/spack/spack/cmd/buildcache.py

```diff
@@ -420,40 +470,52 @@ def push_fn(args):
                 "Use --unsigned to silence this warning."
             )
 
-    # This is a list of installed, non-external specs.
-    specs = bindist.specs_to_be_packaged(
+    specs = _specs_to_be_packaged(
         roots,
-        root="package" in args.things_to_install,
-        dependencies="dependencies" in args.things_to_install,
+        things_to_install=args.things_to_install,
+        build_deps=args.with_build_dependencies or not args.without_build_dependencies,
     )
+
     if not args.private:
         specs = _skip_no_redistribute_for_public(specs)
 
-    # When pushing multiple specs, print the url once ahead of time, as well as how
-    # many specs are being pushed.
     if len(specs) > 1:
         tty.info(f"Selected {len(specs)} specs to push to {push_url}")
 
-    failed = []
+    # Pushing not installed specs is an error. Either fail fast or populate the error list and
+    # push installed package in best effort mode.
+    failed: List[Tuple[Spec, BaseException]] = []
+    with spack.store.STORE.db.read_transaction():
+        if any(not s.installed for s in specs):
+            specs, not_installed = stable_partition(specs, lambda s: s.installed)
+            if args.fail_fast:
+                raise PackagesAreNotInstalledError(not_installed)
+            else:
+                failed.extend(
+                    (s, spack.error.SpackError("package not installed")) for s in not_installed
```

Comment:
Was this meant to be:
```suggestion
                    (s, PackageNotInstalledError("package not installed")) for s in not_installed
```
?